File: src/mcp/src/ramp_mcp_shim/clock.py
"""Clock port for the MCP shim — realises ADR-008 D1 (Python side).

Production policy code MUST consult time through the :class:`Clock`
Protocol rather than calling ``datetime.now`` / ``datetime.utcnow`` /
``time.time`` / ``time.monotonic`` directly. The ruff configuration
under ``src/mcp/pyproject.toml`` bans those primitives inside
``src/mcp/src/ramp_mcp_shim/`` — this module and a small allowlist of
leaves (the production root that constructs the system clock,
observability shims, log formatters) are the only places allowed to
reference the wall clock.

Tests construct :class:`DeterministicClock` and advance it explicitly
so behaviour-gating windows (e.g. RFC 9421 signature ``created``
freshness) are exercised without sleeping.
"""

from __future__ import annotations

import threading
import time as _time_module
from datetime import UTC, datetime, timedelta
from typing import Protocol


class Clock(Protocol):
    """Narrow time port consumed by every behaviour-gating call site.

    Implementations MUST return UTC ``datetime`` instances from :meth:`now`
    and a monotonically non-decreasing float from :meth:`monotonic`. The
    monotonic clock is used for bounded retry / backoff windows where
    wall-clock skew would corrupt the loop.
    """

    def now(self) -> datetime:
        """Return the current instant in UTC."""

    def monotonic(self) -> float:
        """Return a monotonic float reading suitable for elapsed-time math."""


class SystemClock:
    """Production implementation backed by the standard library wall clock.

    This is one of the few sites permitted to call ``datetime.now`` and
    ``time.monotonic`` directly — the ruff allowlist exempts this module.
    """

    def now(self) -> datetime:
        """Return the current UTC instant."""
        return datetime.now(tz=UTC)  # noqa: DTZ005, RUF100 — sole production allowlist

    def monotonic(self) -> float:
        """Delegate to ``time.monotonic``."""
        return _time_module.monotonic()


class DeterministicClock:
    """Test-side implementation. The instant is fixed at construction and
    only advances when the test calls :meth:`advance` or :meth:`set_now`.

    Concurrent reads are safe; concurrent advance from multiple threads is
    serialised by the embedded lock.
    """

    def __init__(self, start: datetime, monotonic_start: float = 0.0) -> None:
        if start.tzinfo is None:
            msg = "DeterministicClock start must be timezone-aware"
            raise ValueError(msg)
        self._lock = threading.Lock()
        self._now = start.astimezone(UTC)
        self._mono = float(monotonic_start)

    def now(self) -> datetime:
        """Return the clock's current UTC instant."""
        with self._lock:
            return self._now

    def monotonic(self) -> float:
        """Return the deterministic monotonic counter."""
        with self._lock:
            return self._mono

    def advance(self, delta: timedelta) -> None:
        """Move the clock forward by ``delta``.

        Negative deltas are clamped to zero so accidental rewinds cannot
        silently break window invariants.
        """
        seconds = max(0.0, delta.total_seconds())
        with self._lock:
            self._now = self._now + timedelta(seconds=seconds)
            self._mono = self._mono + seconds

    def set_now(self, instant: datetime) -> None:
        """Pin the clock to ``instant`` (UTC)."""
        if instant.tzinfo is None:
            msg = "set_now requires timezone-aware datetime"
            raise ValueError(msg)
        with self._lock:
            self._now = instant.astimezone(UTC)


def system_clock() -> Clock:
    """Return the canonical production clock.

    Wiring code that needs a :class:`Clock` for the production path uses
    this factory rather than constructing :class:`SystemClock` inline so
    every wiring site reads the same way and the lint allowlist stays
    narrow to this one module.
    """
    return SystemClock()

File: internal/clock/clock.go
// Package clock realises ADR-008 D1 (time as a port).
//
// Production policy code MUST consult time through the Clock interface
// rather than calling time.Now / time.Since / time.Until directly. The
// forbidigo linter (configured under the policy trees) enforces the
// ban; this package and a small set of allowlisted leaves (production
// roots, observability shims, log formatters) are the only places
// allowed to construct or read the wall clock.
//
// Tests construct DeterministicClock and advance it explicitly so
// behaviour-gating windows (attenuation TTL ≤10m, authority valid_until,
// reporting grace, cache TTL) are exercised without sleeping.
package clock

import (
	"sync"
	"time"
)

// Clock is the canonical time port. Implementations MUST return UTC
// instants; callers rely on UTC semantics for window comparisons. The
// After scheduling primitive is included so deterministic implementations
// can drive scheduled wakeups without real wall-clock latency.
type Clock interface {
	// Now returns the current instant in UTC.
	Now() time.Time
	// After returns a channel that receives a single value once d has
	// elapsed on this clock. Deterministic implementations fire on Advance
	// rather than on real time.
	After(d time.Duration) <-chan time.Time
}

// System is the production implementation; it consults the wall clock.
// This is one of the few sites permitted to call time.Now directly —
// the forbidigo allowlist exempts internal/clock.
type System struct{}

// Now returns the current UTC instant from the wall clock.
func (System) Now() time.Time { return time.Now().UTC() } //nolint:forbidigo // sole production allowlist

// After delegates to the standard library timer.
func (System) After(d time.Duration) <-chan time.Time { return time.After(d) }

// NewSystem returns a System clock value. Provided as a constructor so
// wiring code can write `clock.NewSystem()` symmetrically with
// `clock.NewDeterministic(...)`.
func NewSystem() Clock { return System{} }

// DeterministicClock is the test-side implementation. The instant is
// fixed at construction time and only advances when the test calls
// Advance or SetNow. Concurrent reads are safe; concurrent advance from
// multiple goroutines is serialised by the embedded mutex.
type DeterministicClock struct {
	mu      sync.Mutex
	now     time.Time
	pending []pendingTimer
}

type pendingTimer struct {
	fireAt time.Time
	ch     chan time.Time
}

// NewDeterministic constructs a DeterministicClock anchored at start.
// Tests typically pass a fixed UTC instant such as
// `time.Date(2026, 5, 1, 12, 0, 0, 0, time.UTC)`.
func NewDeterministic(start time.Time) *DeterministicClock {
	return &DeterministicClock{now: start.UTC()}
}

// Now returns the clock's current instant. The returned value is the
// same UTC time as the most recent Advance / SetNow.
func (c *DeterministicClock) Now() time.Time {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.now
}

// Advance moves the clock forward by d and fires any After channels
// whose scheduled instant is now reached. d MUST be non-negative; a
// negative value is treated as zero so accidental rewinds cannot
// silently break window invariants.
func (c *DeterministicClock) Advance(d time.Duration) {
	if d < 0 {
		d = 0
	}
	c.mu.Lock()
	c.now = c.now.Add(d)
	due := c.now
	remaining := c.pending[:0]
	for _, t := range c.pending {
		if !t.fireAt.After(due) {
			t.ch <- due
			close(t.ch)
			continue
		}
		remaining = append(remaining, t)
	}
	c.pending = remaining
	c.mu.Unlock()
}

// SetNow pins the clock to instant; pending After channels whose target
// is now reached fire immediately. Useful when a test needs to jump to
// a specific cliff (e.g. exactly the authority valid_until) rather than
// step forward by a relative duration.
func (c *DeterministicClock) SetNow(instant time.Time) {
	c.mu.Lock()
	c.now = instant.UTC()
	due := c.now
	remaining := c.pending[:0]
	for _, t := range c.pending {
		if !t.fireAt.After(due) {
			t.ch <- due
			close(t.ch)
			continue
		}
		remaining = append(remaining, t)
	}
	c.pending = remaining
	c.mu.Unlock()
}

// After returns a channel that receives once Advance / SetNow brings
// the clock to or past now()+d. Buffered length 1 so Advance never
// blocks even if the test never reads the channel.
func (c *DeterministicClock) After(d time.Duration) <-chan time.Time {
	ch := make(chan time.Time, 1)
	c.mu.Lock()
	defer c.mu.Unlock()
	if d <= 0 {
		ch <- c.now
		close(ch)
		return ch
	}
	c.pending = append(c.pending, pendingTimer{fireAt: c.now.Add(d), ch: ch})
	return ch
}
